README
#####################################################################

    HALKEN vpk0 v0.75.5

    "White Guy That Don't Smile"
    2023/09/12, Tuesday, September 12th; 0715 HOURS

#####################################################################

Preface:

    From: "Pokémon Snap (USA)"
          and
          "NINTENDO Game Boy Advance SDK 3.0/bin/agbcomp.exe"
          and
          "nvpktool_src/nvpktool.cpp"

    "vpk0" decoding functionality was recovered through the act
    of decompiling the Nintendo 64 title, "Pokémon Snap",
    using "Ghidra", a reverse software engineering suite.

    Lempel-Ziv encoding functionality was based upon the routine
    recovered from "agbcomp.exe" using the aforementioned tool,
    "Ghidra", in addition to, "Snowman", an x86 decompiler.

    The bit-stream routine used in the encoding process draws
    significant reference from "nvpktool" by "CaitSith2",
    however, customizable functionality present in their
    source code will not be implemented as it is not within
    the scope for the goal of this project.

#####################################################################

Description:

    This is a tool that handles HAL's "vpk0" data format.

    It's still a work in progress, however, it is a
    functional program.

    * Can decode vpk0 data; both type 0x00 and type 0x01.
    * Encodes data to a bit-stream LZ format that's supported
      by the decoding routine.

#####################################################################

Compiler Flags:
    -std=c99
    -Wall
    -Wextra
    -Wpedantic
    -Werror
    -Os
    -s

    If compiling in a 64-bit environment, use the flags:
    -Wno-int-to-pointer-cast
    -Wno-pointer-to-int-cast

#####################################################################

Motive:

    Nearly 20 years ago, an individual under the alias "CaitSith2"
    successfully reverse-engineered the vpk0 format used by the
    Nintendo e-Reader for the Game Boy Advance.

    This development is certainly no small feat, however, some
    individuals are curious as to whether or not they themselves
    can <fully> reverse-engineer the format focusing on 1:1 accuracy.

    As of February 2022, only the LZ portion of the format has
    been replicated. The Huffman format has yet to be cracked.

#####################################################################


CHANGELOG
.
.   Changelog
.   HAL ヴァリアブル プレフィックス コード 【 vpk 】
.   Research Project
.
----------------------------------------------------------------
@   ==  12 September 2023 [Public]  ==
>   Portions of codebase rewritten.
    Should be easier to read and work with.
!   Code runs slower by a lapse of ~20 seconds.
----------------------------------------------------------------
@   ==  26 September 2022 [Public]  ==
>   Implemented vector-like macro substitutions for the
    conventional functions.
    Using the macros introduces an optimal performance gain
    that yields a 21.31% increase in speed over the
    v0.75.2 release (30/06/2022) with the compiler parameters
    present within the makefile.
----------------------------------------------------------------
@   ==  13 September 2022 [Public]  ==
>   Tweaked the vector-like functions to remove a redundant
    'feature' ( vsetbeg() ).
>   Made some adjustments to the decoding routine for ease
    of reading.
!   Code runs ~1.25% slower....
----------------------------------------------------------------
@   ==  30 June 2022 [Public]  ==
>   Consolidated code into one source file.
>   Implemented vector-like functions, and adjusted code to
    compliment this implementation.
    The code is now roughly 16.47% faster.
!   Condition: Red
    Dunning-Kruger
----------------------------------------------------------------
@   ==  10 May 2022 [Private]  ==
>   Inquisition Stage 3
>   Attempted adjustment for functionality Options.
    Failed formula for automating Sample Modes "0" and "1"
    that would effectively reduce the func. opts. to
    "e = encode" and "d = decode", as opposed to, manually
    entering "s = single", "m = multiple", and "d = decode".
?   Formula theorized the rationale that Mode "1" is
    configured automatically when a percentage of Modulus 4'd
    Offset values are met in proportion to other integral
    data elements.
----------------------------------------------------------------
@   ==  9 May 2022 [Public]  ==
>   Adjusted Encoding Routine to use arbitrary Prefix Codes,
    temporarily.
!   Condition: Orange
    Original Project Incentive is compromised.
----------------------------------------------------------------
@   ==  3 May 2022 [Private]  ==
>   Inquisition Stage 2
>   Manual Alterations to Encoding Routine.
>   Testing of Decoded Sample Data from Super Smash Bros.
    Manually assigning the Tree Headers and Prefix Codes
    for Offset and Length data that were recorded.
----------------------------------------------------------------
@   ==  25 April 2022 [Public]  ==
#   Publicly Released "vpk0" Project under Version 0.75.
!   Condition: Yellow
    Authentic Decoding Routine.
    Speculative/Experimental Encoding Routine.
----------------------------------------------------------------
@   ==  13 January 2022 [Public]  ==
>   Tested the modified routine against 1,512 samples.
!   Condition: Green
    Modified LZ Routine is reproducing extracted LZ Data
    with reliable consistency.
#   Publicly Released Modified LZ Routine with Skeleton Code
    as a separate repository "halkenlz" under Version 1.00.
----------------------------------------------------------------
@   ==  12 January 2022 [Private]  ==
>   LZ Decoding Routine siphoned from "agbcomp.exe".
    Modified to reproduce 1:1 identical HAL LZ binaries.
----------------------------------------------------------------
@   ==  26 December 2021 [Private]  ==
>   Initial Format Inquisition.
>   Decoding Routine siphoned from Pokémon Snap [NPFE].
>   Borrowed CaitSith2's straightforward bitstream LZ
    encoding methodology "nvpktool.exe -level 1", and using
    the SLI LZ algorithm as a experimental component.
----------------------------------------------------------------
.
.   [ end of document ]
.


src/lz.c
/****************************************************************************
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 ***************************************************************************/
 /*--------------------------------------------------------------------------
    HALKEN vpk0 Tool ( Work in Progress ) v0.75.1

    Author  : White Guy That Don't Smile
    Date    : 2022/05/09, Monday, May 9th; 2220 HOURS
    License : UnLicense | Public Domain

    This is a data compression tool presented with the goal of
    *faithfully reverse-engineering the compression aspect of
    HAL Laboratory's enigmatic algorithm.
    It supports decompression by using the code found within
    one of their Nintendo 64 titles, Pokémon Snap.

    As of present, the compression portion is at 50% with the LZ
    routine being faithfully deduced (I hope).

    The exact Huffman routine has yet to be unraveled, therefore,
    this tool temporarily uses bit-stream LZ encoding.

    *This project yields inspiration from CaitSith2 in a number
    of ways, however, the key motive for this project is to
    replicate the format, _faithfully_ (i.e. 1:1 encoded data).
---------------------------------------------------------------------------*/
/*---------------------------------------------------------------------------
    HALKEN LZ Routine v1.00

    Author  : White Guy That Don't Smile
    Date    : 2022/04/25, Monday, April 25th; 2042 HOURS
    License : UnLicense | Public Domain

    This is a Lempel-Ziv compression routine siphoned from the
    AGB SDK tool "agbcomp" - modified to comply with the parameters
    recorded from decoded sample data generated by HAL Laboratory.

    ####  Confirmed LZ Parameters  ####

    Minimum Offset : 0x0001 , Minimum Length : 0x03
    Maximum Offset : 0xFFFF , Maximum Length : 0xFF

    0 Flag : Process 8 bits, verbatim.
    1 Flag : Process Length-Displacement information.

    ####  Supplementary Hypothetical Parameters  ####

    < IF >
        the next reference address, is uneven
        | OR |
        the initial match offset, plus two,
        is not-equal to
        the lookahead match offset
        | OR |
        the lookahead match length minus the initial match length,
        is less-than or equal-to
        three
    < THEN >
        go to the end of the loop
---------------------------------------------------------------------------*/
/*---------------------------------------------------------------------------
    ####  LZ Audits  ####

    <  Nintendo 64  >

        [N64] : Kirby 64: The Crystal Shards
        [???? / ????] : To do....

        [N64] : Pokémon Snap
        [EUR][0003 / 0003] : Success
        [JAP][0003 / 0003] : Success
        [USA][0003 / 0003] : Success

        [N64] : Shigesato Itoi's No. 1 Bass Fishing - Definitive Edition!
        [JAP][0001 / 0001] : Success

        [N64] : Super Smash Bros.
        [EUR][0516 / 0516] : Success
        [JAP][0474 / 0474] : Success
        [USA][0498 / 0498] : Success

    <  Game Boy Advance  >

        [AGB] : e-Reader ROM
        [JAP][0003 / 0003] : Success
        [JAP][0004 / 0004] : Success [e-Reader+]
        [USA][0003 / 0003] : Success

        [AGB] : Pokémon Emerald Version
        [0001 / 0001] : Success

        [AGB] : Pokémon FireRed Version
        [0001 / 0001] : Success

        [AGB] : Pokémon LeafGreen Version
        [0001 / 0001] : Success

        [AGB] : Super Mario Advance 4 - Super Mario Bros. 3
        [0001 / 0001] : Success
---------------------------------------------------------------------------*/
#include "vpk0.h"
#include <stdio.h>
#include <stdlib.h>



static void _search( u8 *src,
                     const u32  srcpos, const u32  srclen,
                           u32 *match_offset, u32 *match_length )
{
  u8 *head;
  u8 *tail;
  u32 offset;
  u32 length;
  u32 window;
  int nominal;

  *match_offset = 0;
  *match_length = 0;
  offset   = 1;
  window   = (srcpos < 0x10000) ? 1 : srcpos - 0xFFFF;
  nominal  = ((srcpos < srclen) << 2) | 2;
  nominal |= window <= srcpos;

  while ( nominal == 7 )
  {
    nominal = 4;
    tail    = &src[((i32)srcpos - offset)];

    if ( *tail == src[srcpos] )
    {
      length = 1;
      head = &src[srcpos];

      while ( (*++head == *++tail) && (++length < 0xFF) );

      if ( srclen <= (srcpos + length) ) length = srclen - srcpos;

      if ( (2 < length) && (*match_length < length) )
      {
        *match_offset = offset;
        *match_length = length;
      }
    }

    nominal |= ((++offset < 0x10000U) << 1);
    nominal |= ++window <= srcpos;
  }

  return;
}



static void _null_block( u8 *def )
{
  if ( def != (u8 *)0 )
  {
    free( def );
    def = (u8 *)0;
  }

  return;
}



static u8 *_produce_lz( u8 *src, u8 **dst, const u32 srclen, u32 **dstlen )
{
  u8  *def;
  u32  dp;
  u32  cp;
  u32  match_offset;
  u32  match_length;
  u32  lookahead_offset;
  u32  lookahead_length;
  u32  mask;
  register u32 srcpos;
  u32  ndp;
  u32  i;
  u8   cmds;
  u8   bytes[3];
  int  nominal;

  cp  = 0;
  dp  = cp + 1U;
  ndp = 0x1000U << ENC_GREED;
  def = malloc( (ndp << 2U) );
  srcpos = 0;
  cmds   = 0;
  mask   = 0x80U;

#define extend_buffer \
  if ( ndp == dp ) \
  { \
    ndp += ((0x1000U << 2) << ENC_GREED); \
\
    if ( (def = realloc( def, ndp )) == (u8 *)0 ) goto err; \
  }

#define shift_mask \
  if ( (mask >>= 1) == 0 ) \
  { \
    mask = 0x80U; \
    def[cp] = cmds; \
    cp = dp++; \
\
    if ( ndp == dp ) \
    { \
      ndp += ((0x1000U << 2) << ENC_GREED); \
\
      if ( (def = realloc( def, ndp )) == (u8 *)0 ) goto err; \
    } \
\
    cmds = 0; \
  }

  while ( srcpos < srclen )
  {
    _search( src, srcpos, srclen, &match_offset, &match_length );

    if ( match_length < 3U )
    {
      def[dp++] = src[srcpos++];
      extend_buffer;
    }
    else
    {
      _search( src, (srcpos + 1U), srclen,
               &lookahead_offset, &lookahead_length );

      if ( (match_length + 1U) <= lookahead_length )
      {
        nominal   = 0;
        def[dp++] = src[srcpos++];
        extend_buffer;
        shift_mask;
        /*---------------------------------------------
        Testing against these conditions seems to fall
        in line with the method needed to reproduce the
        LZ data siphoned from preexisting "vpk0" data.
        ---------------------------------------------*/
        nominal |= (srcpos & 1);
        nominal |= (match_offset + 2U) != lookahead_offset;
        nominal |= (lookahead_length - match_length) <= 3;

        if ( nominal != 0 ) continue;
        else
        {
          match_offset = lookahead_offset;
          match_length = lookahead_length;
        }
      }

      cmds |= mask;
      bytes[0] = (u8)match_offset;
      bytes[1] = (u8)(match_offset >> 8);
      bytes[2] = (u8)match_length;
      i = 0;

      while ( i < 3U )
      {
        def[dp++] = bytes[i++];
        extend_buffer;
      }

      srcpos += match_length;
    }

    shift_mask;
  }

  if ( mask != 0x80U ) def[cp] = cmds;
  else --dp;

  **dstlen = (u32)dp;

  if ( (*dst = (u8 *)calloc( **dstlen, sizeof(u8) )) == (u8 *)0 ) goto err;

  for ( srcpos = 0; srcpos < dp; *(*dst + srcpos) = def[srcpos], srcpos++ );

  _null_block( def );
  return *dst;

err:

  printf("RAM Unavailable!\n");

  if ( *dst != (u8 *)0 ) free( *dst );

  _null_block( def );
  return *dst = (u8 *)0;
}



u8 *_lz( u8 *src, u8 **dst, const u32 srclen, u32 *dstlen )
{
  return *dst = _produce_lz( src, &*dst, srclen, &dstlen );
}


src/utils.c
/****************************************************************************
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 ***************************************************************************/
 /*--------------------------------------------------------------------------
    HALKEN vpk0 Tool ( Work in Progress ) v0.75.1

    Author  : White Guy That Don't Smile
    Date    : 2022/05/09, Monday, May 9th; 2220 HOURS
    License : UnLicense | Public Domain

    This is a data compression tool presented with the goal of
    *faithfully reverse-engineering the compression aspect of
    HAL Laboratory's enigmatic algorithm.
    It supports decompression by using the code found within
    one of their Nintendo 64 titles, Pokémon Snap.

    As of present, the compression portion is at 50% with the LZ
    routine being faithfully deduced (I hope).

    The exact Huffman routine has yet to be unraveled, therefore,
    this tool temporarily uses bit-stream LZ encoding.

    *This project yields inspiration from CaitSith2 in a number
    of ways, however, the key motive for this project is to
    replicate the format, _faithfully_ (i.e. 1:1 encoded data).
---------------------------------------------------------------------------*/
#include "vpk0.h"

u16 _swap16( u16 data )
{
  return (u16)(((data & 0x00FF) << 8) |
               ((data & 0xFF00) >> 8));
}

u32 _swap32( u32 data )
{
  return (u32)(((u32)_swap16((u16)(data & 0x0000FFFF)) << 16) |
               ((u32)_swap16((u16)((data & 0xFFFF0000) >> 16))));
}


include/vpk0.h
/****************************************************************************
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 ***************************************************************************/
 /*--------------------------------------------------------------------------
    HALKEN vpk0 Tool ( Work in Progress ) v0.75.1

    Author  : White Guy That Don't Smile
    Date    : 2022/05/09, Monday, May 9th; 2220 HOURS
    License : UnLicense | Public Domain

    This is a data compression tool presented with the goal of
    *faithfully reverse-engineering the compression aspect of
    HAL Laboratory's enigmatic algorithm.
    It supports decompression by using the code found within
    one of their Nintendo 64 titles, Pokémon Snap.

    As of present, the compression portion is at 50% with the LZ
    routine being faithfully deduced (I hope).

    The exact Huffman routine has yet to be unraveled, therefore,
    this tool temporarily uses bit-stream LZ encoding.

    *This project yields inspiration from CaitSith2 in a number
    of ways, however, the key motive for this project is to
    replicate the format, _faithfully_ (i.e. 1:1 encoded data).
---------------------------------------------------------------------------*/
#ifndef LIBSLI_H_INCLUDED
#define LIBSLI_H_INCLUDED



#define VPK 0x76706B30



/*------------------------------------------------------
At the expense of RAM, adjusting this macro value
will improve execution time during encoding by reducing
the frequency of re-allocation calls with the work
buffer(s) in use, in addition to, increasing the
possibility of errors concerning integer overflows or
allocation rejection.
The aforementioned macro corresponds to the number of
bits that will be left-shifted by a hard-coded value
as part of an incremental supplement in conjunction
with work RAM buffers used for storing encoded data
representation elements.
[Safe]
  0 =   4096 (Standard)
  1 =   8192
[Risky]
  2 =  16384
  3 =  32768
[Asking for Trouble]
  4 =  65536
  5 = 131072
[Absolute Madman]
  6 = 262144
  7 = 524288
------------------------------------------------------*/
#define ENC_GREED 2



typedef   signed char  i8;
typedef unsigned char  u8;
typedef   signed short i16;
typedef unsigned short u16;
#ifdef __x86_64__
typedef   signed int   i32;
typedef unsigned int   u32;
#else
typedef   signed long  i32;
typedef unsigned long  u32;
#endif



u16 _swap16( u16 data );
u32 _swap32( u32 data );
u8 *_lz( u8 *src, u8 **dst, const u32 srclen, u32 *dstlen );
void encode( u8 *src, u8 **dst,
             const u32 srclen, u32 *dstlen,
             const u32 sample_depth );
void decode( const u8 *src, u8 *dst, const u32 decsz );



#endif


src/encode.c
/****************************************************************************
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 ***************************************************************************/
 /*--------------------------------------------------------------------------
    HALKEN vpk0 Tool ( Work in Progress ) v0.75.1

    Author  : White Guy That Don't Smile
    Date    : 2022/05/09, Monday, May 9th; 2220 HOURS
    License : UnLicense | Public Domain

    This is a data compression tool presented with the goal of
    *faithfully reverse-engineering the compression aspect of
    HAL Laboratory's enigmatic algorithm.
    It supports decompression by using the code found within
    one of their Nintendo 64 titles, Pokémon Snap.

    As of present, the compression portion is at 50% with the LZ
    routine being faithfully deduced (I hope).

    The exact Huffman routine has yet to be unraveled, therefore,
    this tool temporarily uses bit-stream LZ encoding.

    *This project yields inspiration from CaitSith2 in a number
    of ways, however, the key motive for this project is to
    replicate the format, _faithfully_ (i.e. 1:1 encoded data).
---------------------------------------------------------------------------*/
/*---------------------------------------------------------------------------
    Things that need to be done:
 >  Need to implement Huffman coding.
    For now, this routine borrows CaitSith2's "-level 1" for functionality.
    Once the Huffman aspect has been implemented, this functionality
    will be rescinded for conformance.
 >  Need to verify that writing 16 bits at a time is a legal operation.
    The decoding routine from the N64 loads this breadth amount at a time,
    however, this is likely invalid with hardware like the e-Reader.
 >  Need to review the 0x01 Method conditional routine.
    As of present, it is based on both the reference from CaitSith2's
    solution, as well as, the decoding routine from Pokémon Snap.
    While the code for this tool can decode the encoded 0x01 data
    it generates, CaitSith2's "nvpktool" is unable to decode the
    aforementioned data.
    It's a conflicting observation.
---------------------------------------------------------------------------*/
#include "vpk0.h"
#include <stdio.h>
#include <stdlib.h>



static u32 _nbits = 16U;
static u16 _frame = 0;



/*----------------------------------------------
    Regard clause #2 in the ToDo list as to,
    whether or not, this function needs to
    write 32-bit breadths at a time.
----------------------------------------------*/
static void _store_bits( u16 *buf, u32 *cursor,
                         u16 data, u32 breadth )
{
  u32 i;

  for ( i = 0; i < breadth; i++ )
  {
    _frame |= (((data >> (breadth - 1 - i)) & 1) << --_nbits);

    if ( _nbits == 0 )
    {
      *(buf + *cursor) = _swap16( _frame );
      (*cursor)++;
      _nbits = 16;
      _frame = 0;
    }
  }

  return;
}



static u32 _get_breadth( u32 data )
{
  u32 breadth;

  for ( breadth = 0; data > 0; breadth++ ) data >>= 1;

  return breadth;
}



void encode( u8 *src, u8 **dst,
             const u32 isize, u32 *osize,
             const u32 sample_depth )
{
  i32  flags  = 0;
  u8  *retval = (u8  *)0;
  u32  lz_pos = 0;
  u32  nflags = 0;
  u32  cursor = 0;
  u32  cursor_boundary = 0x1000U << ENC_GREED;
  u16 *bit_stream = (u16 *)calloc( cursor_boundary , sizeof(u16) );
  u32  breadth;
  u32  div;
  u32  and;

#define extend_buffer \
  if ( cursor == cursor_boundary ) \
  { \
    cursor_boundary += (0x1000U << ENC_GREED); \
    bit_stream = realloc( bit_stream, (cursor_boundary * 2U) ); \
\
    if ( bit_stream == (u16 *)0 ) goto err; \
  }

  if ( bit_stream == (u16 *)0 ) goto err;

  _store_bits( bit_stream, &cursor, (u16)(VPK   >> 16), 16 );
  _store_bits( bit_stream, &cursor, (u16)(VPK        ), 16 );
  _store_bits( bit_stream, &cursor, (u16)(isize >> 16), 16 );
  _store_bits( bit_stream, &cursor, (u16)(isize      ), 16 );
  _store_bits( bit_stream, &cursor, (u16)sample_depth,   8 );
  retval = _lz( src, &*dst, isize, &*osize );

  if ( retval == (u8 *)0 ) goto err;
  else
  {
    u16 bit;

    /*_store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor, 16, 8 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  8, 8 );
    _store_bits( bit_stream, &cursor,  1, 1 );*/

    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  1, 8 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  3, 8 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor, 10, 8 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor, 16, 8 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  1, 1 );

    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  2, 8 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  4, 8 );
    _store_bits( bit_stream, &cursor,  0, 1 );
    _store_bits( bit_stream, &cursor,  8, 8 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  1, 1 );
    _store_bits( bit_stream, &cursor,  1, 1 );

    while ( lz_pos < *osize )
    {
      if ( nflags == 0 )
      {
        flags   = *(*dst + lz_pos++);
        flags <<= 24;
        nflags  = 8;
      }
      else
      {
        bit = flags < 0;
        _store_bits( bit_stream, &cursor, bit, 1 );
        extend_buffer;

        if ( bit )
        {
          if ( !sample_depth )
          {
            breadth = _get_breadth( (u32)*(u16 *)(*dst + lz_pos) );

            if ( breadth <= 1 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x0, 1 );
              breadth = 1;
            }
            else if ( breadth <= 3 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x2, 2 );
              breadth = 3;
            }
            else if ( breadth <= 10 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x6, 3 );
              breadth = 10;
            }
            else
            {
              _store_bits( bit_stream, &cursor, (u16)0x7, 3 );
              breadth = 16;
            }

            extend_buffer;
            _store_bits( bit_stream, &cursor, *(u16 *)(*dst + lz_pos), breadth );
            extend_buffer;

            /*_store_bits( bit_stream, &cursor, *(u16 *)(*dst + lz_pos), 16 );
            extend_buffer;*/
          }
          else
          {
            div = ((*(u16 *)(*dst + lz_pos) + 8) / 4);
            and = (*(u16 *)(*dst + lz_pos) & 3);

            if ( and )
            {
              breadth = _get_breadth( (and - 1) );

              if ( breadth <= 1 )
              {
                _store_bits( bit_stream, &cursor, (u16)0x0, 1 );
                breadth = 1;
              }
              else
              {
                _store_bits( bit_stream, &cursor, (u16)0x2, 2 );
                breadth = 3;
              }

              extend_buffer;
              _store_bits( bit_stream, &cursor, (u16)(and - 1), breadth );
              extend_buffer;
            }

            breadth = _get_breadth( div );

            if ( breadth <= 1 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x0, 1 );
              breadth = 1;
            }
            else if ( breadth <= 3 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x2, 2 );
              breadth = 3;
            }
            else if ( breadth <= 10 )
            {
              _store_bits( bit_stream, &cursor, (u16)0x6, 3 );
              breadth = 10;
            }
            else
            {
              _store_bits( bit_stream, &cursor, (u16)0x7, 3 );
              breadth = 16;
            }

            extend_buffer;
            _store_bits( bit_stream, &cursor, (u16)div, breadth );
            extend_buffer;

            /*if ( (*(u16 *)(*dst + lz_pos) % 4) == 0 )
            {
              _store_bits( bit_stream,
                           &cursor,
                           (u16)((*(u16 *)(*dst + lz_pos) + 8) / 4),
                           16 );
              extend_buffer;
            }
            else
            {
              _store_bits( bit_stream,
                           &cursor,
                           (u16)((*(u16 *)(*dst + lz_pos) % 4) + -1),
                           16 );
              extend_buffer;
              _store_bits( bit_stream,
                           &cursor,
                           (u16)((*(u16 *)(*dst + lz_pos) + 8) / 4),
                           16 );
              extend_buffer;
            }*/
          }

          lz_pos += 2;
          breadth = _get_breadth( (u32)*(*dst + lz_pos) );

          if ( breadth <= 2 )
          {
            _store_bits( bit_stream, &cursor, (u16)0x0, 1 );
            breadth = 2;
          }
          else if ( breadth <= 4 )
          {
            _store_bits( bit_stream, &cursor, (u16)0x2, 2 );
            breadth = 4;
          }
          else
          {
            _store_bits( bit_stream, &cursor, (u16)0x3, 2 );
            breadth = 8;
          }

          extend_buffer;
          _store_bits( bit_stream, &cursor, (u16)*(*dst + lz_pos++), breadth );
          extend_buffer;

          /*_store_bits( bit_stream, &cursor, (u16)*(*dst + lz_pos++), 8 );
          extend_buffer;*/
        }
        else
        {
          _store_bits( bit_stream, &cursor, (u16)*(*dst + lz_pos++), 8 );
          extend_buffer;
        }

        flags <<= 1;
        nflags--;
      }
    }

    if ( _nbits != 16 )
    {
      _store_bits( bit_stream, &cursor, 0, _nbits );
      extend_buffer;
    }

    *osize = cursor << 1;
    *dst = (u8 *)bit_stream;
  }

  return;

err:

  printf( "Unsuccessful process!\n" );

  if ( retval != (u8 *)0 )
  {
    free( retval );
    retval = (u8 *)0;
  }

  if ( bit_stream != (u16 *)0 )
  {
    free( bit_stream );
    bit_stream = (u16 *)0;
  }

  if ( *dst != (u8 *)0 )
  {
    free( *dst );
    *dst = (u8 *)0;
  }

  *osize = 0;

  return;
}
